### PR TITLE
Fix some "attr-defined" mypy errors: Explicitly re-export relied on a…

### PR DESCRIPTION
**Pull request by adamtheturtle**
…ttributes

This removes errors in the format:

```
tests/test_codeblock.py:10: error: Module "sybil.parsers.codeblock" does not explicitly export attribute "CodeBlockParser"  [attr-defined]
```

when running:

```
mypy --strict sybil tests
```

### CODE REVIEW

File: sybil/__init__.py [hunk 1]
```diff
@@ -2,3 +2,12 @@
...
 from .example import Example
+
+__all__ = [
+    'Sybil',
+    'Document',
+    'Region',
+    'LexedRegion',
+    'Lexeme',
+    'Example',
+]
```
Comment (cjw296) [inline sybil/__init__.py:13]: Why is this needed?

**Review by cjw296 — COMMENTED** [review]
This PR makes me pretty sad: it would appear that mypy is forcing a bunch of code duplication here?

**Comment by adamtheturtle** [comment]
Thank you for the review @cjw296 .

If you don't like this, we can instead set the `mypy` configuration option [`implicit_reexport`](https://mypy.readthedocs.io/en/stable/config_file.html#confval-implicit_reexport).

I tend to think that this approach is better, because I consider it an example of "Explicit is better than implicit". For example, if I have the following files:

```python
# foo.py

def foo_fn():
  pass
```

```python
# bar.py

from foo import foo_fn

def bar_fn():
  foo_fn()
```

```python
# baz.py

from bar import foo_fn

def baz_fn():
  oo_fn()
```

ofen this is done by accident - the author should have imported `foo_fn` from `foo` as the author of `bar.py` wrote `bar.py` without expecting that `bar.foo_fn` is a public interface, and something that would be imported from `bar`. This `mypy` check provides that safety from relying on unexpected public interfaces from imports, at least for code with this type check. In this project, it seems that you only want folks to rely on being able to import from imports in a few files - this `__all__` method is a way of saying that "these are the few files, and few imports that I allow people to rely on in that way".

As a (minor?) benefit, doing it this way allows you to know that any unused imports can be removed without changing the interface you explicitly provide. You can then use tooling to remove unusued imports. For example, Ruff [implements an auto-fix](https://beta.ruff.rs/docs/rules/unused-import/) for Flake8's [F401](https://www.flake8rules.com/rules/F401.html), removing unused imports.

In this project, I can see that there are a few unused imports:

```
sybil/parsers/abstract/clear.py:1:30: F401 [*] `typing.Optional` imported but unused
sybil/parsers/myst/clear.py:1:8: F401 [*] `re` imported but unused
sybil/parsers/rest/clear.py:1:8: F401 [*] `re` imported but unused
sybil/sybil.py:4:77: F401 [*] `typing.List` imported but unused
sybil/typing.py:1:65: F401 [*] `typing.Any` imported but unused
```

In my own projects, I run `ruff` on save/pre-commit and never have to manually remove an import again.
Of course, you could tell flake8/ruff to ignore particular files, but that, to me, seems equivalent to `__all__`.

Please let me know how you would like for me to proceed.

**Comment by cjw296** [comment]
Yeah, I think the "make sure imports come from the right place" is compelling enough to be worth the code duplication.
PyCharm has a habit of auto-importing from bad locations, hopefully this will help!
